### PR TITLE
Disable cosmetic filtering on Speedreader pages

### PR DESCRIPTION
**Pull request by boocmp**
Resolves https://github.com/brave/brave-browser/issues/40642

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/views/brave_actions/brave_shields_action_view.cc [hunk 1]
```diff
@@ -217,11 +223,41 @@ void BraveShieldsActionView::ButtonPressed() {
...
+#if BUILDFLAG(ENABLE_SPEEDREADER)
+  if (auto* speedreader_tab_helper =
+          speedreader::SpeedreaderTabHelper::FromWebContents(web_contents)) {
+    return speedreader::DistillStates::IsDistilled(
+        speedreader_tab_helper->PageDistillState());
+  }
+#endif
+  return false;
+}
+bool BraveShieldsActionView::ShouldShowBubble(
```
Comment (iefremov) [inline browser/ui/views/brave_actions/brave_shields_action_view.cc:240]: whitespace

File: browser/speedreader/speedreader_tab_helper.cc [hunk 2]
```diff
@@ -355,6 +356,18 @@ void SpeedreaderTabHelper::UpdateUI() {
...
+void SpeedreaderTabHelper::ReadyToCommitNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInPrimaryMainFrame()) {
+    return;
+  }
+
+  blink::web_pref::WebPreferences prefs =
+      web_contents()->GetOrCreateWebPreferences();
+  prefs.page_in_reader_mode = DistillStates::IsDistilled(PageDistillState());
+  web_contents()->SetWebPreferences(prefs);
```
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.cc:368]: i'm slightly concerned with the fact that the vast majority of navigations are not into reader mode, so we will do lots of useless IPC. Can we trigger this only when the pref is actually changed?

**Comment by boocmp** [comment]
@ShivanKaul Yes, you are right. The actual disabling of shields for the specified tab will require quite a lot of changes, because now the settings work at the ContentSettings level. In general, this could be useful, but so far there has been no need to disable shields for a specific tab (AFAIK). Disabling CF is enough to solve problems with the reading mode I would even prefer not to disable the shields button.

**Comment by ShivanKaul** [comment]
@boocmp @iefremov We do have issues with Speedreader + network blocking too: https://bravesoftware.slack.com/archives/CLX7EV42Z/p1726603890011589?thread_ts=1726595972.638809&cid=CLX7EV42Z

**Comment by ShivanKaul** [comment]
Spoke with @boocmp and I understand now that network blocking is not relevant here, at least for the [issue I mentioned](https://bravesoftware.slack.com/archives/CLX7EV42Z/p1726603890011589?thread_ts=1726595972.638809&cid=CLX7EV42Z). Let's go with just CF disabling.

**Review by iefremov — APPROVED** [review]

**Review by ShivanKaul — APPROVED** [review]
Thanks!

**Comment by ShivanKaul** [comment]
(also just FYI @antonok-edm)

**Pull request merged**